File: macos-app/Lexis/Services/ElevenLabsClient.swift
import Foundation

struct VoiceTuning: Sendable {
    var stability: Double
    var similarityBoost: Double
    var style: Double
    var speed: Double
}

struct Voice: Sendable, Identifiable, Hashable {
    var id: String
    var name: String
}

protocol SpeechService: Sendable {
    func synthesize(text: String, voiceID: String, model: String, settings: VoiceTuning) async throws -> (audio: Data, timings: [WordTiming])
    func voices() async throws -> [Voice]
}

/// Talks to the ElevenLabs REST API directly (no SDK).
struct ElevenLabsClient: SpeechService {
    private let secrets: SecretsStore

    init(secrets: SecretsStore = KeychainStore()) {
        self.secrets = secrets
    }

    func synthesize(text: String, voiceID: String, model: String, settings: VoiceTuning) async throws -> (audio: Data, timings: [WordTiming]) {
        guard let apiKey = secrets.get(.elevenLabsAPIKey), !apiKey.isEmpty else {
            throw LexisError.missingAPIKey(service: "ElevenLabs")
        }
        guard let url = URL(string: "https://api.elevenlabs.io/v1/text-to-speech/\(voiceID)/with-timestamps?output_format=mp3_44100_128") else {
            throw LexisError.decodingFailure(service: "ElevenLabs", underlying: "invalid voice ID")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "xi-api-key")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "text": text,
            "model_id": model,
            "voice_settings": [
                "stability": settings.stability,
                "similarity_boost": settings.similarityBoost,
                "style": settings.style,
                "speed": settings.speed,
                "use_speaker_boost": true,
            ],
        ])

        let data = try await RetryPolicy.withRetry(serviceName: "ElevenLabs") {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw LexisError.httpFailure(service: "ElevenLabs", status: -1, body: "")
            }
            guard (200...299).contains(http.statusCode) else {
                if http.statusCode == 429 { throw RetryableError.rateLimited }
                let bodyText = String(data: data, encoding: .utf8) ?? ""
                throw LexisError.httpFailure(service: "ElevenLabs", status: http.statusCode, body: bodyText)
            }
            return data
        }

        struct Alignment: Decodable {
            let characters: [String]
            let character_start_times_seconds: [Double]
            let character_end_times_seconds: [Double]
        }
        struct Response: Decodable {
            let audio_base64: String
            let alignment: Alignment?
        }

        let decoded: Response
        do {
            decoded = try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw LexisError.decodingFailure(service: "ElevenLabs", underlying: error.localizedDescription)
        }
        guard let audioData = Data(base64Encoded: decoded.audio_base64) else {
            throw LexisError.decodingFailure(service: "ElevenLabs", underlying: "invalid base64 audio")
        }

        let timings = Self.charsToWords(
            characters: decoded.alignment?.characters ?? [],
            starts: decoded.alignment?.character_start_times_seconds ?? [],
            ends: decoded.alignment?.character_end_times_seconds ?? []
        )
        return (audioData, timings)
    }

    func voices() async throws -> [Voice] {
        guard let apiKey = secrets.get(.elevenLabsAPIKey), !apiKey.isEmpty else {
            throw LexisError.missingAPIKey(service: "ElevenLabs")
        }
        var request = URLRequest(url: URL(string: "https://api.elevenlabs.io/v1/voices")!)
        request.httpMethod = "GET"
        request.setValue(apiKey, forHTTPHeaderField: "xi-api-key")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            let bodyText = String(data: data, encoding: .utf8) ?? ""
            throw LexisError.httpFailure(service: "ElevenLabs", status: status, body: bodyText)
        }

        struct VoicesResponse: Decodable {
            struct VoiceEntry: Decodable {
                let voice_id: String
                let name: String
            }
            let voices: [VoiceEntry]
        }
        do {
            let decoded = try JSONDecoder().decode(VoicesResponse.self, from: data)
            return decoded.voices.map { Voice(id: $0.voice_id, name: $0.name) }
        } catch {
            throw LexisError.decodingFailure(service: "ElevenLabs", underlying: error.localizedDescription)
        }
    }

    /// Ports backend/api/utils.py's _chars_to_words verbatim: any whitespace
    /// is a boundary; a word's start is its first character's start time, its
    /// end is its last character's end time.
    static func charsToWords(characters: [String], starts: [Double], ends: [Double]) -> [WordTiming] {
        var words: [WordTiming] = []
        var currentChars: [String] = []
        var currentStart: Double?
        var currentEnd: Double = 0

        let count = min(characters.count, starts.count, ends.count)
        for index in 0..<count {
            let char = characters[index]
            let isWhitespace = char.rangeOfCharacter(from: .whitespacesAndNewlines) != nil
            if isWhitespace {
                if !currentChars.isEmpty {
                    words.append(WordTiming(word: currentChars.joined(), start: currentStart ?? 0, end: currentEnd))
                    currentChars = []
                    currentStart = nil
                }
            } else {
                if currentStart == nil { currentStart = starts[index] }
                currentChars.append(char)
                currentEnd = ends[index]
            }
        }
        if !currentChars.isEmpty {
            words.append(WordTiming(word: currentChars.joined(), start: currentStart ?? 0, end: currentEnd))
        }
        return words
    }
}

File: android-app/app/src/main/java/com/aboooooo57/lexume/network/ElevenLabsClient.kt
package com.aboooooo57.lexume.network

import android.util.Base64
import com.aboooooo57.lexume.data.local.SecretKey
import com.aboooooo57.lexume.data.local.SecureKeyStore
import com.aboooooo57.lexume.data.model.WordTiming
import com.aboooooo57.lexume.support.LexumeException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/** Mirrors `Services/ElevenLabsClient.swift`'s `VoiceTuning`. */
data class VoiceTuning(val stability: Double, val similarityBoost: Double, val style: Double, val speed: Double)

/** Mirrors `Services/ElevenLabsClient.swift`'s `Voice`. */
data class Voice(val id: String, val name: String)

/** Result of a synthesis call: audio bytes + per-word timings. */
data class SynthesisResult(val audioData: ByteArray, val timings: List<WordTiming>)

/** Mirrors `Services/ElevenLabsClient.swift`'s `SpeechService` protocol. */
interface SpeechService {
    suspend fun synthesize(text: String, voiceId: String, model: String, settings: VoiceTuning): SynthesisResult
    suspend fun voices(): List<Voice>
}

/** Talks to the ElevenLabs REST API directly (no SDK). Mirrors `Services/ElevenLabsClient.swift`. */
class ElevenLabsClient(private val secureKeyStore: SecureKeyStore) : SpeechService {
    override suspend fun synthesize(text: String, voiceId: String, model: String, settings: VoiceTuning): SynthesisResult {
        val apiKey = secureKeyStore.get(SecretKey.ELEVENLABS_API_KEY)
        if (apiKey.isNullOrEmpty()) throw LexumeException.MissingApiKey("ElevenLabs")

        val body = JSONObject()
            .put("text", text)
            .put("model_id", model)
            .put(
                "voice_settings",
                JSONObject()
                    .put("stability", settings.stability)
                    .put("similarity_boost", settings.similarityBoost)
                    .put("style", settings.style)
                    .put("speed", settings.speed)
                    .put("use_speaker_boost", true)
            )

        val responseBody = RetryPolicy.withRetry("ElevenLabs") {
            val request = Request.Builder()
                .url("https://api.elevenlabs.io/v1/text-to-speech/$voiceId/with-timestamps?output_format=mp3_44100_128")
                .addHeader("Content-Type", "application/json")
                .addHeader("xi-api-key", apiKey)
                .post(body.toString().toRequestBody(JSON_MEDIA_TYPE))
                .build()

            val (statusCode, text) = HttpClients.shared.executeSuspending(request)
            if (statusCode == 429) throw RateLimitedException()
            if (statusCode !in 200..299) {
                throw LexumeException.HttpFailure("ElevenLabs", statusCode, text.take(200))
            }
            text
        }

        val json = try {
            JSONObject(responseBody)
        } catch (e: Exception) {
            throw LexumeException.DecodingFailure("ElevenLabs", e.message ?: "malformed response")
        }
        val audioData = try {
            Base64.decode(json.getString("audio_base64"), Base64.DEFAULT)
        } catch (e: Exception) {
            throw LexumeException.DecodingFailure("ElevenLabs", "invalid base64 audio")
        }

        val alignment = json.optJSONObject("alignment")
        val characters = alignment?.optJSONArray("characters")
        val starts = alignment?.optJSONArray("character_start_times_seconds")
        val ends = alignment?.optJSONArray("character_end_times_seconds")
        val timings = charsToWords(
            characters = (0 until (characters?.length() ?: 0)).map { characters!!.getString(it) },
            starts = (0 until (starts?.length() ?: 0)).map { starts!!.getDouble(it) },
            ends = (0 until (ends?.length() ?: 0)).map { ends!!.getDouble(it) }
        )
        return SynthesisResult(audioData, timings)
    }

    override suspend fun voices(): List<Voice> {
        val apiKey = secureKeyStore.get(SecretKey.ELEVENLABS_API_KEY)
        if (apiKey.isNullOrEmpty()) throw LexumeException.MissingApiKey("ElevenLabs")

        val request = Request.Builder()
            .url("https://api.elevenlabs.io/v1/voices")
            .get()
            .addHeader("xi-api-key", apiKey)
            .build()
        val (statusCode, responseBody) = HttpClients.shared.executeSuspending(request)
        if (statusCode !in 200..299) {
            throw LexumeException.HttpFailure("ElevenLabs", statusCode, responseBody.take(200))
        }

        return try {
            val voicesArray = JSONObject(responseBody).getJSONArray("voices")
            (0 until voicesArray.length()).map { i ->
                val entry = voicesArray.getJSONObject(i)
                Voice(id = entry.getString("voice_id"), name = entry.getString("name"))
            }
        } catch (e: Exception) {
            throw LexumeException.DecodingFailure("ElevenLabs", e.message ?: "malformed response")
        }
    }

    private companion object {
        val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}

/**
 * Ports `backend/api/utils.py`'s `_chars_to_words` (via
 * `ElevenLabsClient.swift`'s `charsToWords`) verbatim: any whitespace is a
 * boundary; a word's start is its first character's start time, its end is
 * its last character's end time.
 */
fun charsToWords(characters: List<String>, starts: List<Double>, ends: List<Double>): List<WordTiming> {
    val words = mutableListOf<WordTiming>()
    val currentChars = StringBuilder()
    var currentStart: Double? = null
    var currentEnd = 0.0

    val count = minOf(characters.size, starts.size, ends.size)
    for (index in 0 until count) {
        val char = characters[index]
        val isWhitespace = char.isNotEmpty() && char.all { it.isWhitespace() }
        if (isWhitespace) {
            if (currentChars.isNotEmpty()) {
                words.add(WordTiming(currentChars.toString(), currentStart ?: 0.0, currentEnd))
                currentChars.clear()
                currentStart = null
            }
        } else {
            if (currentStart == null) currentStart = starts[index]
            currentChars.append(char)
            currentEnd = ends[index]
        }
    }
    if (currentChars.isNotEmpty()) {
        words.add(WordTiming(currentChars.toString(), currentStart ?: 0.0, currentEnd))
    }
    return words
}
